File: golang/next_greater_element.go
package main

import "fmt"

/*
 * golang这个语言比较特殊
 * golang不提供deque数据结构,
 * golang 切片只提供append(a, x)追加接口, 删除末尾元素用s[:len(s)-1]
 *
 * golang range一个列表, 返回ids和value
 * golang map返回key，可以同时返回key是否存在。if v, ok := d[x]; ok {
 */

func nextGreaterElement(nums1 []int, nums2 []int) (ans []int) {
	stk := []int{}
	d := map[int]int{}
	for i := len(nums2) - 1; i >= 0; i-- {
		x := nums2[i]
		for len(stk) > 0 && stk[len(stk)-1] < x {
			stk = stk[:len(stk)-1]
		}
		if len(stk) > 0 {
			d[x] = stk[len(stk)-1]
		}
		stk = append(stk, x)
	}
	for _, x := range nums1 {
		if v, ok := d[x]; ok {
			ans = append(ans, v)
		} else {
			ans = append(ans, -1)
		}
	}
	return
}

func main() {
	nums1 := []int{4, 1, 2}
	nums2 := []int{1, 3, 4, 2}
	fmt.Println(nextGreaterElement(nums1, nums2))
}

File: java/NextGreaterElement.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;
import java.util.HashMap;

/*
 * java Deque的定义, Deque<Integer> stk = new ArrayDeque<>()
 * Deque的常用函数, push(x), pop(), peek()
 * 
 * java Map的定义, Map<Integer, Integer> xx = new HashMap<>()
 * Map的常用函数, containsKey(x)判断key存在, get(x) 根据key获得value, put(x,y) 插入key,value对
 * getOrDefault(x, y) 如果key存在, 返回value, 否则返回y
 */

class NextGreaterElement {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Deque<Integer> stk = new ArrayDeque<>();
        int m = nums1.length, n = nums2.length;
        Map<Integer, Integer> d = new HashMap<>(n);

        for (int i = n-1; i >= 0; --i) {
            int x = nums2[i];
            while (!stk.isEmpty() && stk.peek() < x) {
                stk.pop();
            }
            if (!stk.isEmpty()) {
                d.put(x, stk.peek());
            }
            stk.push(x);
        }

        int[] ans = new int[m];
        for (int i = 0; i < m; ++i) {
            ans[i] = d.getOrDefault(nums1[i], -1);
        }
        return ans;
    }

    public static void main(String[] args) {
        NextGreaterElement nge = new NextGreaterElement();
        int[] ret = nge.nextGreaterElement(new int[]{4,1,2}, new int[]{1,3,4,2});
        for (int i = 0; i < ret.length; i++) {
            System.out.println(ret[i]);
        }
    }
}